src/Storages/StorageFile.cpp

```diff
@@ -526,12 +495,182 @@ namespace
+                        format,
+                        archive);
+                }
+
+                auto archive_reader = createArchiveReader(archive);
+
+                auto try_get_columns_from_schema_cache = [&](const std::string & full_path) -> std::optional<ColumnsDescription>
+                {
+                    auto context = getContext();
+                    if (!getContext()->getSettingsRef().schema_inference_use_cache_for_file)
+                        return std::nullopt;
+
+                    auto & schema_cache = StorageFile::getSchemaCache(context);
+                    auto get_last_mod_time = [&]() -> std::optional<time_t>
+                    {
+                        if (0 != stat(archive_reader->getPath().c_str(), &file_stat))
+                            return std::nullopt;
+
+                        return file_stat.st_mtime;
+                    };
+
+                    auto cache_key = getKeyForSchemaCache(full_path, format, format_settings, context);
+                    auto columns = schema_cache.tryGetColumns(cache_key, get_last_mod_time);
+
+                    if (columns)
+                        return columns;
+
+                    return std::nullopt;
+                };
+
+                if (archive_info.isSingleFileRead())
+                {
+                    read_buf = archive_reader->readFile(archive_info.path_in_archive, false);
+                    ++current_archive_index;
+                    if (!read_buf)
+                        continue;
+
+                    last_read_file_path = processed_files.emplace_back(fmt::format("{}::{}", archive_reader->getPath(), archive_info.path_in_archive));
+                    columns_from_cache = try_get_columns_from_schema_cache(last_read_file_path);
+
+                    if (columns_from_cache)
+                        return nullptr;
+                }
+                else
+                {
+                    auto file_enumerator = archive_reader->firstFile();
+                    if (!file_enumerator)
+                    {
+                        if (getContext()->getSettingsRef().engine_file_skip_empty_files)
+                        {
+                            read_files_from_archive.clear();
+                            ++current_archive_index;
+                            continue;
+                        }
+
+                        throw Exception(
+                            ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE,
+                            "Cannot extract table structure from {} format file, because the archive {} has no files. "
+                            "You must specify table structure manually",
+                            format,
+                            archive);
+                    }
+
+                    const auto * filename = &file_enumerator->getFileName();
+                    while (read_files_from_archive.contains(*filename) || !archive_info.filter(*filename))
```

Comment:
Hmm, so each next() call we create archive reader and iterate over all files until we found not processed yet using unordered set? Looks really inefficient, especially when archive contains a lot of files (they can be small, so we need to read a lot of files for schema inference). Why don't we create archive reader and file_enumerator once per archive and then reuse them to iterate through files?

Reply:
yes, it's really strange and inefficient but it's because we need the current active read buffer to get file_enumerator back (internally a handler is passed around between those objects).